File: python/robot-name/robot_name.py
import random
import string


class Robot(object):
    names = set()

    def __init__(self):
        while True:
            self.name = '{0}{1}{2}{3}{4}'.format(self.generate_letter(), self.generate_letter(),
                                                 self.generate_digit(), self.generate_digit(), self.generate_digit())
            if self.name not in self.names:
                self.names.add(self.name)
                break

    def reset(self):
        self.__init__()

    @classmethod
    def generate_letter(cls):
        return random.choice(string.ascii_uppercase)

    @classmethod
    def generate_digit(cls):
        return random.randrange(10)

File: csharp/robot-name/RobotName.cs
using System.Collections.Generic;
using System;

public class Robot
{
    const int MAX_NAME_NUM = 26 * 26 * 10 * 10 * 10;

    static Random random = new Random();
    static ISet<string> names = new HashSet<string>();

    public Robot()
    {
        Reset();
    }

    public string Name { get; private set; }

    public void Reset()
    {
        if (names.Count == MAX_NAME_NUM)
        {
            throw new InvalidOperationException();
        }

        while (true)
        {
            string name =
                $"{GenerateLetter()}{GenerateLetter()}{GenerateDigit()}{GenerateDigit()}{GenerateDigit()}";
            if (!names.Contains(name))
            {
                names.Add(name);
                Name = name;

                break;
            }
        }
    }

    char GenerateLetter() => (char)(random.Next(26) + 'A');

    int GenerateDigit() => random.Next(10);
}
